live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/adapters/auth/AuthenticationService.java

```diff
@@ -0,0 +1,177 @@
+    }
+
+    public void authenticate(boolean forceAuthentication)
+    {
+        if (forceAuthentication || securityContextIsEmpty() || tokenHasOrIsAboutToExpire(AUTH_SCHEDULE_DELAY_MINUTES))
+        {
+            String clientName = oAuth2ClientProperties.getRegistration().get(CLIENT_REGISTRATION_ID).getClientName();
+            OAuth2AuthenticationToken authenticationToken = createOAuth2AuthenticationToken(clientName);
+            Authentication authentication = authenticationManager.authenticate(authenticationToken);
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+    }
+
+    public static void setAuthorizationToken(Exchange exchange)
+    {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        if (authentication instanceof OAuth2LoginAuthenticationToken authenticationToken)
+        {
+            OAuth2AccessToken accessToken = authenticationToken.getAccessToken();
+            Map<String, Object> principalAttributes = authenticationToken.getPrincipal().getAttributes();
+
+            String authorization = accessToken.getTokenType().getValue() + " " + accessToken.getTokenValue();
+            exchange.getIn().setHeaders(Map.of(
+                    AUTHORIZATION, authorization,
+                    ENVIRONMENT_KEY_HEADER, principalAttributes.get(ENVIRONMENT_KEY_ATTRIBUTE_KEY)));
+        }
+        else
+        {
+            throw new LiveIngesterRuntimeException("Spring security context does not contain authentication principal of type " + OAuth2LoginAuthenticationToken.class.getSimpleName());
+        }
+    }
+
+    private boolean securityContextIsEmpty()
+    {
+        return SecurityContextHolder.getContext() == null
+                || SecurityContextHolder.getContext().getAuthentication() == null;
+    }
+
+    private boolean tokenHasOrIsAboutToExpire(int scheduleDelayMinutes)
+    {
+        return !(SecurityContextHolder.getContext().getAuthentication() instanceof OAuth2LoginAuthenticationToken authenticationToken)
+                || authenticationToken.getAccessToken() == null
+                || authenticationToken.getAccessToken().getExpiresAt() == null
+                || tokenHasOrIsAboutToExpire(authenticationToken.getAccessToken(), scheduleDelayMinutes);
+    }
+
+    private boolean tokenHasOrIsAboutToExpire(OAuth2AccessToken accessToken, int scheduleDelayMinutes)
+    {
+        if (Optional.ofNullable(accessToken.getIssuedAt()).isEmpty() || Optional.ofNullable(accessToken.getExpiresAt()).isEmpty())
+        {
+            return true;
+        }
+
+        Duration expirationOffset = Duration.between(accessToken.getIssuedAt(), accessToken.getExpiresAt()).minusMinutes(scheduleDelayMinutes);
+        long expirationOffsetSeconds = expirationOffset.isNegative() ? 0 : expirationOffset.toSeconds();
+        return Instant.now().isAfter(accessToken.getExpiresAt().minusSeconds(expirationOffsetSeconds));
```

Comment:
I didn't understand the code that is meant to handle "or is about to expire". I didn't think we would need something like this anyway because we planned to use a period of 55 minutes for a 60 minute token.

Reply:
Ah - is the idea that you're checking if the expiry will be before the next execution of this task?

Reply:
While testing I had a lot of calls earlier than 55 minutes form token creation and this check was preventing the app from generating a new one when it wasn't necessary 🤔 I'm not sure if it will be useful though, so I can remove it

Reply:
I'm thinking about running this logic every 55 min or in case of "access denied exception", or 401 🤔

Reply:
Yes - I assumed we would try to authenticate every 55 minutes and that way we'd ensure we always had at least one (sometimes two) tokens that would work.